### DISCUSSION TITLE
Assert equality on collections: can the error message describe the difference between the two collections?

### DISCUSSION THREAD

**Discussion by prjm** [xunit/xunit#2149] [General]
I am somewhat unhappy with the exception text for the equality assertion of collections.
For example:

```
    public class UnitTest1
    {
        [Fact]
        public void Test1()
        {
            var expected = new string[30];
            var result = new string[30];
            result[29] = "X";
            Assert.Equal(expected, result);
        }
    }
```
I get the following error message (within Visual Studio):

>  XUnitTestProject1.UnitTest1.Test1
>    Source: UnitTest1.cs line 8
>    Duration: 10 ms
> 
>   Message: 
>     Assert.Equal() Failure
>     Expected: String[] [null, null, null, null, null, ...]
>     Actual:   String[] [null, null, null, null, null, ...]
>   Stack Trace: 
>     UnitTest1.Test1() line 13

With that message, it is hard to see the difference between the two collections. Can I improve the message myself or should I write a customized assert method? For simple cases, I'd like to get a clue on the difference of the two collections...

**Comment by bradwilson** [comment]
This was recently fixed: https://github.com/xunit/xunit/commit/f4be18e8bf926b3ec0ead036fca5df7fb2114116?w=1

Example:

```csharp
[Fact]
public void ExactArraySize_DifferenceNearStart()
{
	var expectedMessage =
		"Assert.Equal() Failure" + Environment.NewLine +
		"              ↓ (pos 1)" + Environment.NewLine +
		"Expected: [1, 2, 3, 4, 5]" + Environment.NewLine +
		"Actual:   [1, 99, 3, 4, 5]" + Environment.NewLine +
		"              ↑ (pos 1)";

	var ex = Record.Exception(
		() => Assert.Equal(
			new List<int> { 1, 2, 3, 4, 5 },
			new List<int> { 1, 99, 3, 4, 5 }
		)
	);

	Assert.NotNull(ex);
	Assert.Equal<object>(expectedMessage, ex.Message);
}
```

**Reply by prjm** [reply]
That's nice to hear! Thank you for your reply.
